### ISSUE TITLE
CA1031 should ignore "safe" methods

### ISSUE THREAD

**Issue by CoolDadTx** [dotnet/roslyn-analyzers#3211]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v2.9.8 (Latest)

#### Diagnostic ID

CA1031

#### Repro steps

Note: this is strict a simple example, not something you might actually write.

```csharp
int TryConvert ( string value )
{
   try
   {
      return Int32.Parse(value);
   } catch (Exception e)
   {   
   };
   return 0;
}
```
1. Compile the above code and you, correctly, get the warning that you shouldn't catch generic exception

#### Expected behavior

The issue with this warning is that there are cases where methods are "safe". Safe methods are methods that are attempting to do something and the expectation is that they can fail but that it isn't an error that should be bubbled up. 

Let's ignore for purposes of this discussion that some exceptions (like `OutOfMemoryException`) cannot be correctly handled by the exception. Because C# doesn't have an easy way to say "handle all exceptions except - other than exception filters) it is difficult to properly build a series of try-catch blocks that would correctly handle all the exceptions you might care about and the few you don't, short of an exception filter. Since the purpose of this warning is to just let you know something might be wrong it makes sense to be able to ignore it when it doesn't make sense. 

The problem is that the existing suppression syntax (outside a global suppression) is either a verbose code analysis attribute or 2 pragmas. It would be nice if certain methods could be considered "safe" out of the box and therefore be ignored by this warning. Off the top of my head I'd argue that `Safe-` and `Try-` methods would qualify but I could easily see other methods that might as well. I'd like to see either these 2 types of methods auto-excluded and/or a special attribute be added (or perhaps there is one already) such that a method can be marked safe for this type of error. Again, the existing analysis attribute just seems too wordy and the pragma clutters up things. I believe this was done for the validating null argument rule so something similar would be nice here.

#### Actual behavior

Currently you have to suppress this warning in "safe" methods even though most of the time such exceptions are silently logged/eaten/etc.

**Comment by Evangelink** [comment]
I don't think we should hardcode some "safe" method pattern but we could maybe add the [excluded_symbol_names](https://github.com/dotnet/roslyn-analyzers/blob/master/docs/Analyzer%20Configuration.md#excluded-symbol-names) option for this rule with the tweak suggested in #3242

In the meantime if you think that the pragma and attributes are too cumbersome, what about implementing some kind of a generic method like `void SafeInvoke(Action action)` and its pendant `T SafeInvoke<T>(Func<T> func)`? This way you have only 2 places where to do this general try/catch/log pattern?

**Comment by CoolDadTx** [comment]
Provided we can exclude `Try` and `Safe` methods I'm fine with any implementation.

I do have a safe invoke method but the problem with this approach is that it requires an extra delegate allocation over simply calling the method directly. For high performance code that is called a lot this is an issue. For now I'm just using the attribute everywhere. It isn't pretty but it clarifies why we're doing it.

**Comment by mavasani** [comment]
I tend to agree with @Evangelink here. We should not hard code anything about special patterns in the rule here, but just provide a configurable editorconfig option, possibly with wildcards, to skip this analyzer on specific parts of code.

**Comment by Evangelink** [comment]
@mavasani Do you see any better way than having the excluded_symbol_names option supporting wildcard and being applied here?

**Comment by mavasani** [comment]
Sounds reasonable.

**Comment by Evangelink** [comment]
Hum before starting the change, is there any mechanism in place to be able to measure analyzers performance just to ensure I am not killing perfs by allowing the wildcard on the `excluded_symbol_names` option.

**Comment by mavasani** [comment]
@Evangelink I am presuming we would check the option only after we have identified there is a violation, so it should be fine to perform wildcard processing then.

**Comment by Evangelink** [comment]
I will give it a go and we will see if that's ok or not :)
